File: UniCalib/unicalib/extrinsic/fine/ikalibr_wrapper.py
"""
iKalibr 封装器 —— 连续时间 B-spline 多传感器联合精标定
项目位置: /root/calib_ws/src/iKalibr/
"""
from __future__ import annotations
import logging
import os
import subprocess
import tempfile
from pathlib import Path
from typing import Dict, Optional, Tuple

import numpy as np
import yaml
from scipy.spatial.transform import Rotation

from ...core.calib_result import CalibResult
from ...core.sensor_config import SensorConfig, SensorType
from ...core.data_manager import DataManager

logger = logging.getLogger(__name__)

WORKSPACE = os.environ.get("WORKSPACE", "/root/calib_ws")
IKALIBR_WS = WORKSPACE  # iKalibr 编译在 colcon 工作空间中


class IKalibrWrapper:
    """
    iKalibr 连续时间 B-spline 多传感器联合精标定封装。

    核心能力:
      - IMU-LiDAR 外参 (R, t) + 时间偏移 td
      - IMU-Camera 外参 (通过 LiDAR 桥接)
      - B-spline 轨迹表示 (SO3 × R3)
      - Ceres 非线性优化

    调用 iKalibr ROS2 节点:
      ros2 run ikalibr ikalibr_solver --ros-args -p config_path:=<config>
    """

    def __init__(self, system_config: dict):
        cfg = (system_config.get("calibration", {})
               .get("stage3_fine", {})
               .get("global_optimization", {}))
        self.spline_order = cfg.get("spline_order", 4)
        self.knot_so3 = cfg.get("knot_distance_so3", 0.02)
        self.knot_pos = cfg.get("knot_distance_pos", 0.02)
        self.max_iter = cfg.get("max_iterations", 50)
        self.enable_td = cfg.get("enable_time_offset", True)
        self.robust_kernel = cfg.get("robust_kernel", "cauchy")
        self.robust_param = cfg.get("robust_kernel_param", 1.0)
        self._available = self._check_ikalibr()

    def joint_optimize(
        self,
        data_mgr: DataManager,
        sensors: Dict[str, SensorConfig],
        intrinsics: Dict,
        coarse_results: Dict[Tuple[str, str], CalibResult],
    ) -> Dict[Tuple[str, str], CalibResult]:
        """
        全局联合 B-spline 优化所有传感器外参。
        返回精化后的外参字典。
        """
        if not self._available:
            logger.warning("iKalibr not available, returning coarse results.")
            return coarse_results

        try:
            with tempfile.TemporaryDirectory() as tmpdir:
                config = self._build_config(
                    sensors, intrinsics, coarse_results, tmpdir, data_mgr)
                config_path = os.path.join(tmpdir, "ikalibr_joint.yaml")
                with open(config_path, "w") as f:
                    yaml.dump(config, f, default_flow_style=False)

                logger.info(f"  Running iKalibr joint optimization...")
                logger.debug(f"  Config: {config_path}")

                env = os.environ.copy()
                proc = subprocess.run(
                    [
                        "ros2", "run", "ikalibr", "ikalibr_prog",
                        "--ros-args",
                        "-p", f"config_path:={config_path}",
                    ],
                    env=env,
                    capture_output=True, text=True,
                    timeout=1800,  # 30 分钟超时
                )

                if proc.returncode != 0:
                    logger.error(f"iKalibr failed:\n{proc.stderr[-500:]}")
                    return coarse_results

                # 解析结果
                result_dir = os.path.join(tmpdir, "result")
                if os.path.exists(result_dir):
                    return self._parse_results(result_dir, coarse_results)
                else:
                    logger.warning("iKalibr result directory not found.")
                    return coarse_results

        except subprocess.TimeoutExpired:
            logger.error("iKalibr timed out (1800s).")
        except Exception as e:
            logger.error(f"iKalibr error: {e}", exc_info=True)

        return coarse_results

    # ------------------------------------------------------------------
    # 配置文件构建
    # ------------------------------------------------------------------

    def _build_config(
        self,
        sensors: Dict[str, SensorConfig],
        intrinsics: Dict,
        coarse_results: Dict[Tuple[str, str], CalibResult],
        output_dir: str,
        data_mgr: DataManager,
    ) -> dict:
        """构建 iKalibr 联合优化配置文件。"""

        config = {
            "Calibration": {
                "Type": "MULTI_SENSOR",
                "OutputPath": os.path.join(output_dir, "result"),
                "DataPath": str(data_mgr.data_path),

                # B-Spline 配置
                "BSpline": {
                    "SplineOrder": self.spline_order,
                    "KnotTimeDistance": {
                        "SO3": self.knot_so3,
                        "Pos": self.knot_pos,
                    },
                },

                # 优化配置
                "Optimization": {
                    "MaxIterations": self.max_iter,
                    "EnableTimeOffset": self.enable_td,
                    "TimeOffsetPadding": 0.1,  # ±100ms 搜索范围
                    "RobustKernel": self.robust_kernel,
                    "RobustKernelParam": self.robust_param,
                },

                # 传感器配置
                "Sensors": {},

                # 初始外参
                "InitialExtrinsics": {},
            }
        }

        for sid, sensor in sensors.items():
            sensor_cfg: dict = {
                "Type": self._sensor_type_str(sensor.sensor_type),
                "Topic": sensor.topic,
                "FrameId": sensor.frame_id,
            }

            if sensor.is_camera():
                intr = intrinsics.get(sid)
                if intr and hasattr(intr, "to_dict"):
                    sensor_cfg["Intrinsic"] = intr.to_dict()
            elif sensor.is_lidar() and sensor.lidar_type:
                sensor_cfg["LidarType"] = sensor.lidar_type.value

            if sensor.rate:
                sensor_cfg["Rate"] = sensor.rate

            config["Calibration"]["Sensors"][sid] = sensor_cfg

        # 填充初始外参
        for (sa, sb), result in coarse_results.items():
            key = f"{sa}_to_{sb}"
            config["Calibration"]["InitialExtrinsics"][key] = {
                "Rotation": result.rotation.tolist(),
                "Translation": result.translation.tolist(),
                "TimeOffset": result.time_offset,
            }

        return config

    # ------------------------------------------------------------------
    # 结果解析
    # ------------------------------------------------------------------

    def _parse_results(
        self, result_dir: str,
        fallback: Dict[Tuple[str, str], CalibResult],
    ) -> Dict[Tuple[str, str], CalibResult]:
        """解析 iKalibr 输出的 YAML 结果文件。"""
        results = dict(fallback)

        result_yaml = os.path.join(result_dir, "calibration_result.yaml")
        if not os.path.exists(result_yaml):
            # 尝试其他命名
            for fname in os.listdir(result_dir):
                if fname.endswith(".yaml"):
                    result_yaml = os.path.join(result_dir, fname)
                    break

        if not os.path.exists(result_yaml):
            logger.warning(f"iKalibr result YAML not found in {result_dir}")
            return results

        with open(result_yaml) as f:
            data = yaml.safe_load(f)

        extrinsics = data.get("Extrinsics", data.get("extrinsics", {}))
        for key_str, ext_data in extrinsics.items():
            parts = key_str.replace("_to_", "|").split("|")
            if len(parts) != 2:
                continue
            sa, sb = parts[0].strip(), parts[1].strip()

            R = np.array(ext_data.get("Rotation", np.eye(3).tolist()))
            t = np.array(ext_data.get("Translation", [0, 0, 0]))
            td = float(ext_data.get("TimeOffset", 0.0))
            reproj = float(ext_data.get("ReprojectionError", float('inf')))

            results[(sa, sb)] = CalibResult(
                pair=(sa, sb),
                rotation=R, translation=t.flatten(),
                time_offset=td,
                reprojection_error=reproj,
                confidence=0.9,
                method_used="iKalibr_bspline",
            )
            logger.info(f"  iKalibr result [{sa}->{sb}]: "
                        f"R_euler={Rotation.from_matrix(R).as_euler('zyx', degrees=True)}, "
                        f"t={t}, td={td*1000:.2f}ms, reproj={reproj:.3f}px")

        return results

    # ------------------------------------------------------------------
    # 辅助
    # ------------------------------------------------------------------

    @staticmethod
    def _sensor_type_str(stype: SensorType) -> str:
        mapping = {
            SensorType.IMU: "SENSOR_IMU",
            SensorType.LIDAR: "SENSOR_LIDAR",
            SensorType.CAMERA_PINHOLE: "SENSOR_CAMERA",
            SensorType.CAMERA_FISHEYE: "SENSOR_CAMERA_FISHEYE",
        }
        return mapping.get(stype, "SENSOR_UNKNOWN")

    @staticmethod
    def _check_ikalibr() -> bool:
        try:
            result = subprocess.run(
                ["ros2", "pkg", "list"], capture_output=True, text=True, timeout=5)
            available = "ikalibr" in result.stdout.lower()
            if not available:
                logger.warning("iKalibr ROS2 package not found.")
            return available
        except Exception:
            return False

File: unicalib_C_plus_plus/src/ikalibr_wrapper.cpp
/**
 * UniCalib C++ — iKalibr 联合精化：构建配置、调用 ros2、解析结果
 */
#include "unicalib/ikalibr_wrapper.hpp"
#include "unicalib/process_runner.hpp"
#include "unicalib/logger.hpp"
#include <filesystem>
#include <fstream>
#include <sstream>

#if defined(UNICALIB_USE_YAML) && UNICALIB_USE_YAML
#include <yaml-cpp/yaml.h>
#endif

namespace fs = std::filesystem;
namespace unicalib {

namespace {

std::string sensor_type_to_ikalibr(SensorType t) {
  switch (t) {
    case SensorType::IMU: return "SENSOR_IMU";
    case SensorType::LIDAR: return "SENSOR_LIDAR";
    case SensorType::CAMERA_PINHOLE: return "SENSOR_CAMERA";
    case SensorType::CAMERA_FISHEYE: return "SENSOR_CAMERA_FISHEYE";
    default: return "SENSOR_UNKNOWN";
  }
}

std::string lidar_type_to_string(LidarType t) {
  switch (t) {
    case LidarType::SPINNING: return "spinning";
    case LidarType::SOLID_STATE: return "solid_state";
    default: return "spinning";
  }
}

}  // namespace

std::optional<std::unordered_map<std::string, CalibResult>> run_ikalibr_joint(
  DataManager& data_mgr,
  const std::string& data_path,
  const std::vector<SensorConfig>& sensors,
  const std::unordered_map<std::string, IntrinsicResultHolder>& intrinsics,
  const std::unordered_map<std::string, CalibResult>& coarse_extrinsics,
  const ExternalToolsConfig& tools_config,
  int timeout_seconds) {

  (void)data_mgr;
  if (tools_config.ikalibr.empty()) {
    LOG_WARNING("iKalibr not configured for joint optimization.");
    LOG_WARNING("Set 'third_party.ikalibr' in config.yaml or set UNICALIB_IKALIBR environment variable.");
    LOG_WARNING("Example config:");
    LOG_WARNING("  third_party:");
    LOG_WARNING("    ikalibr: \"/path/to/iKalibr\"");
    LOG_WARNING("Example environment variable:");
    LOG_WARNING("  export UNICALIB_IKALIBR=/path/to/iKalibr");
    LOG_WARNING("Note: iKalibr requires ROS2 environment. Ensure ROS_DOMAIN_ID is set.");
    return std::nullopt;
  }

#if !(defined(UNICALIB_USE_YAML) && UNICALIB_USE_YAML)
  LOG_WARNING("iKalibr requires yaml-cpp");
  return std::nullopt;
#else
  std::string tmpdir = (fs::temp_directory_path() / "unicalib_ikalibr").string();
  try {
    fs::create_directories(tmpdir);
  } catch (...) {
    LOG_ERROR("iKalibr: failed to create " + tmpdir);
    return std::nullopt;
  }

  std::string config_path = tmpdir + "/ikalibr_config.yaml";
  std::string result_dir = tmpdir + "/result";

  // --- 构建 Calibration 配置 ---
  YAML::Emitter out;
  out << YAML::BeginMap;
  out << YAML::Key << "Calibration" << YAML::Value << YAML::BeginMap;
  out << YAML::Key << "Type" << YAML::Value << "MULTI_SENSOR";
  out << YAML::Key << "OutputPath" << YAML::Value << result_dir;
  out << YAML::Key << "DataPath" << YAML::Value << data_path;
  out << YAML::Key << "BSpline" << YAML::Value << YAML::BeginMap;
  out << YAML::Key << "SplineOrder" << YAML::Value << 4;
  out << YAML::Key << "KnotTimeDistance" << YAML::Value << YAML::BeginMap;
  out << YAML::Key << "SO3" << YAML::Value << 0.2;
  out << YAML::Key << "Pos" << YAML::Value << 0.2;
  out << YAML::EndMap << YAML::EndMap;
  out << YAML::Key << "Optimization" << YAML::Value << YAML::BeginMap;
  out << YAML::Key << "MaxIterations" << YAML::Value << 200;
  out << YAML::Key << "EnableTimeOffset" << YAML::Value << true;
  out << YAML::Key << "TimeOffsetPadding" << YAML::Value << 0.1;
  out << YAML::Key << "RobustKernel" << YAML::Value << true;
  out << YAML::Key << "RobustKernelParam" << YAML::Value << 1.0;
  out << YAML::EndMap;

  out << YAML::Key << "Sensors" << YAML::Value << YAML::BeginMap;
  for (const auto& s : sensors) {
    out << YAML::Key << s.sensor_id << YAML::Value << YAML::BeginMap;
    out << YAML::Key << "Type" << YAML::Value << sensor_type_to_ikalibr(s.sensor_type);
    out << YAML::Key << "Topic" << YAML::Value << s.topic;
    out << YAML::Key << "FrameId" << YAML::Value << s.frame_id;

    auto it_i = intrinsics.find(s.sensor_id);
    if (s.is_camera() && it_i != intrinsics.end()) {
      if (it_i->second.camera_pinhole) {
        const auto& K = it_i->second.camera_pinhole->K;
        const auto& d = it_i->second.camera_pinhole->dist_coeffs;
        out << YAML::Key << "Intrinsic" << YAML::Value << YAML::BeginMap;
        out << YAML::Key << "K" << YAML::Value << YAML::Flow << YAML::BeginSeq;
        for (int i = 0; i < 3; ++i) for (int j = 0; j < 3; ++j) out << K(i, j);
        out << YAML::EndSeq;
        out << YAML::Key << "dist_coeffs" << YAML::Value << YAML::Flow << YAML::BeginSeq;
        for (double c : d) out << c;
        out << YAML::EndSeq << YAML::EndMap;
      } else if (it_i->second.camera_fisheye) {
        out << YAML::Key << "Intrinsic" << YAML::Value << YAML::BeginMap;
        out << YAML::Key << "model_type" << YAML::Value << it_i->second.camera_fisheye->model_type;
        for (const auto& [k, v] : it_i->second.camera_fisheye->params)
          out << YAML::Key << k << YAML::Value << v;
        out << YAML::EndMap;
      }
    } else if (s.is_lidar() && s.lidar_type) {
      out << YAML::Key << "LidarType" << YAML::Value << lidar_type_to_string(*s.lidar_type);
    }

    if (s.rate) out << YAML::Key << "Rate" << YAML::Value << *s.rate;
    out << YAML::EndMap;
  }
  out << YAML::EndMap;

  out << YAML::Key << "InitialExtrinsics" << YAML::Value << YAML::BeginMap;
  for (const auto& [key, r] : coarse_extrinsics) {
    std::string ext_key = r.pair.first + "_to_" + r.pair.second;
    out << YAML::Key << ext_key << YAML::Value << YAML::BeginMap;
    out << YAML::Key << "Rotation" << YAML::Value << YAML::BeginSeq;
    for (int i = 0; i < 3; ++i)
      out << YAML::Flow << YAML::BeginSeq << r.rotation(i,0) << r.rotation(i,1) << r.rotation(i,2) << YAML::EndSeq;
    out << YAML::EndSeq;
    out << YAML::Key << "Translation" << YAML::Value << YAML::Flow << YAML::BeginSeq
        << r.translation(0) << r.translation(1) << r.translation(2) << YAML::EndSeq;
    out << YAML::Key << "TimeOffset" << YAML::Value << r.time_offset;
    out << YAML::EndMap;
  }
  out << YAML::EndMap << YAML::EndMap << YAML::EndMap;  // InitialExtrinsics, Calibration, root

  try {
    std::ofstream f(config_path);
    if (!f) { LOG_ERROR("iKalibr: cannot write " + config_path); return std::nullopt; }
    f << out.c_str();
  } catch (const std::exception& e) {
    LOG_ERROR(std::string("iKalibr config write: ") + e.what());
    return std::nullopt;
  }

  // --- 调用 ros2 run ikalibr ikalibr_prog ---
  std::string abs_config = fs::absolute(config_path).string();
  std::vector<std::string> argv = {
    "ros2", "run", "ikalibr", "ikalibr_prog",
    "--ros-args", "-p", "config_path:=" + abs_config
  };
  ProcessResult res = run_process(argv, {}, timeout_seconds, "");
  if (res.exit_code != 0 || res.timed_out) {
    LOG_WARNING("iKalibr process exit_code=" + std::to_string(res.exit_code)
                + " timed_out=" + (res.timed_out ? "1" : "0"));
    if (!res.stderr_output.empty()) LOG_WARNING("iKalibr stderr: " + res.stderr_output.substr(0, 500));
    return std::nullopt;
  }

  // --- 解析 result_dir/calibration_result.yaml ---
  fs::create_directories(result_dir);
  std::string result_yaml = result_dir + "/calibration_result.yaml";
  if (!fs::exists(result_yaml)) {
    for (const auto& e : fs::directory_iterator(result_dir)) {
      if (e.path().extension() == ".yaml") {
        result_yaml = e.path().string();
        break;
      }
    }
  }
  if (!fs::exists(result_yaml)) {
    LOG_WARNING("iKalibr result YAML not found in " + result_dir);
    return std::nullopt;
  }

  YAML::Node data;
  try {
    data = YAML::LoadFile(result_yaml);
  } catch (const std::exception& e) {
    LOG_ERROR(std::string("iKalibr result parse: ") + e.what());
    return std::nullopt;
  }

  auto extr = data["Extrinsics"] ? data["Extrinsics"] : data["extrinsics"];
  std::unordered_map<std::string, CalibResult> results;
  if (!extr || !extr.IsMap()) return std::nullopt;

  for (auto it = extr.begin(); it != extr.end(); ++it) {
    std::string key_str = it->first.as<std::string>();
    const YAML::Node& ext_data = it->second;
    // key_str 形如 "cam_to_lidar" 或 "sensor_a_to_sensor_b"
    std::string delim = "_to_";
    size_t pos = key_str.find(delim);
    if (pos == std::string::npos) continue;
    std::string sa = key_str.substr(0, pos);
    std::string sb = key_str.substr(pos + delim.size());
    std::string our_key = sa + ":" + sb;

    Eigen::Matrix3d R = Eigen::Matrix3d::Identity();
    Eigen::Vector3d t(0, 0, 0);
    double td = 0.0, reproj = std::numeric_limits<double>::infinity();
    if (ext_data["Rotation"]) {
      const auto& Rnode = ext_data["Rotation"];
      for (int i = 0; i < 3 && i < static_cast<int>(Rnode.size()); ++i)
        for (int j = 0; j < 3 && j < static_cast<int>(Rnode[i].size()); ++j)
          R(i, j) = Rnode[i][j].as<double>();
    }
    if (ext_data["Translation"]) {
      const auto& tnode = ext_data["Translation"];
      for (size_t i = 0; i < 3 && i < tnode.size(); ++i) t(i) = tnode[i].as<double>();
    }
    if (ext_data["TimeOffset"]) td = ext_data["TimeOffset"].as<double>();
    if (ext_data["ReprojectionError"]) reproj = ext_data["ReprojectionError"].as<double>();

    CalibResult cr;
    cr.pair = {sa, sb};
    cr.rotation = R;
    cr.translation = t;
    cr.time_offset = td;
    cr.reprojection_error = reproj;
    cr.confidence = 0.9;
    cr.method_used = "iKalibr_bspline";
    results[our_key] = std::move(cr);
    LOG_INFO("iKalibr result [" + sa + "->" + sb + "] td=" + std::to_string(td*1000) + "ms reproj=" + std::to_string(reproj));
  }

  return results;
#endif
}

}  // namespace unicalib
